src/main/java/com/spotify/futures/CompletableFutures.java

```diff
@@ -156,6 +160,38 @@ private CompletableFutures() {
     return collectingAndThen(toList(), CompletableFutures::allAsList);
   }
 
+  /**
+   * Apply an asynchronous operation to every entity in a stream, and collect all replies into a
+   * map.
+   *
+   * <p> Usage:
+   *
+   * <pre>{@code
+   *   collection.stream()
+   *      .collect(joinMap(this::someAsyncFunc))
+   *      .thenApply(this::consumeMap)
+   * }</pre>
+   *
+   * <p> The generated {@link CompletableFuture} will complete to a map from entity to the result
+   * of the supplied asynchronous opoertaion.  Similar to
+   * {@link CompletableFuture#allOf(CompletableFuture[])}, if any of the map operations complete
+   * exceptionally, then the returned CompletableFuture also does so, with a
+   * {@link CompletionException} holding this exception as its cause.
+   *
+   * @param <K> the key type in the output map
+   * @param <V> the value type in the output map
+   * @return a new {@link CompletableFuture} according to the rules outlined in the method
+   * description
+   * @since 0.3.3
+   */
+  public static <K, V> Collector<K, ?, CompletableFuture<Map<K, V>>> joinMap(
```

Comment:
Random thought. also add a function similar to allAsList:
`CompletionStage<Map<K, V>> allAsList(Map<K, CompletionStage<V>> input);`
That could be useful if the construction of the futures is different than a single transform on the keys

Reply:
Sure, a separate PR makes sense!

On a related note, it might be worth to port this function too: https://github.com/spotify/futures-extra/blob/master/src/main/java/com/spotify/futures/FuturesExtra.java#L609
It seems like a more generalized version, and you could build the allAsMap using that.